cpp/include/Ice/LocalException.h

```diff
@@ -25,21 +25,173 @@
+            : LocalException(file, line),
+              id(std::move(id)),
+              facet(std::move(facet)),
+              operation(std::move(operation))
+        {
+        }
+
+        /**
+         * Construct a RequestFailedException without specifying the details of the current request. The details will
+         * be filled-in automatically by the Ice runtime before marshaling the exception.
+         * @param file The file name in which the exception was raised, typically __FILE__.
+         * @param line The line number at which the exception was raised, typically __LINE__.
+         */
+        RequestFailedException(const char* file, int line) noexcept : LocalException(file, line) {}
+
+        void ice_print(std::ostream& stream) const override;
+
+        /**
+         * The identity of the Ice Object to which the request was sent.
+         */
+        const Identity id;
+
+        /**
+         * The facet to which the request was sent.
+         */
+        const std::string facet;
+
+        /**
+         * The operation name of the request.
+         */
+        const std::string operation;
+    };
+
+    /**
+     * This exception is raised if an object does not exist on the server, that is, if no facets with the given identity
+     * exist.
+     * \headerfile Ice/Ice.h
+     */
+    class ICE_API ObjectNotExistException final : public RequestFailedException
+    {
+    public:
+        using RequestFailedException::RequestFailedException;
+
+        const char* ice_id() const noexcept final;
     };
 
+    /**
+     * This exception is raised if no facet with the given name exists, but at least one facet with the given identity
+     * exists.
+     * \headerfile Ice/Ice.h
+     */
+    class ICE_API FacetNotExistException final : public RequestFailedException
+    {
+    public:
+        using RequestFailedException::RequestFailedException;
+
+        const char* ice_id() const noexcept final;
+    };
+
+    /**
+     * This exception is raised if an operation for a given object does not exist on the server. Typically this is
+     * caused by either the client or the server using an outdated Slice specification.
+     * \headerfile Ice/Ice.h
+     */
+    class ICE_API OperationNotExistException final : public RequestFailedException
+    {
+    public:
+        using RequestFailedException::RequestFailedException;
+
+        const char* ice_id() const noexcept final;
+    };
+
+    /**
+     * This exception is raised if an operation call on a server raises an unknown exception. For example, for C++, this
+     * exception is raised if the server throws a C++ exception that is not directly or indirectly derived from
+     * <code>Ice::LocalException</code> or <code>Ice::UserException</code>.
+     * \headerfile Ice/Ice.h
+     */
+    class ICE_API UnknownException : public LocalException
+    {
+    public:
+        using LocalException::LocalException;
+
+        const char* ice_id() const noexcept override;
```

Comment:
We could make _what protected and provide a public reason as an alias for _what:
```
const char* const reason;
```

but since the type changes (from string), it's not that convenient.